python/mlc_llm/support/preshard.py

```diff
@@ -101,35 +108,57 @@ def _compile_shard_funcs(mod: IRModule, device: Device):
 
 
 def apply_preshard(
-    quantize_map: Any,
     named_params: Dict[str, nn.Parameter],
     tensor_parallel_shards: int,
     args: Any,
-):
-    """Update quantize_map and named_params, create shard functions based on shard strategies."""
-    model_config = args.model.config.from_file(args.config)
-    model_config.tensor_parallel_shards = tensor_parallel_shards
-    model = args.model.model(model_config)
-    model.to(args.quantization.model_dtype)
-
+) -> Tuple[Dict[str, nn.Parameter], Dict[str, Callable[[NDArray], Sequence[NDArray]]]]:
+    """Apply pre-sharding to the named parameters.
+
+    Parameters
+    ----------
+    named_params : Dict[str, nn.Parameter]
+        The named parameters of the model. If the model is quantized, the named parameters should
+        the state dictionary of the quantized model.
+    tensor_parallel_shards : int
+        The number of tensor parallel shards.
+    args : Any
+        The parsed arguments of weight conversion.
+
+    Returns
+    -------
+    Tuple[Dict[str, nn.Parameter], Dict[str, Callable[[NDArray], Sequence[NDArray]]]
+        The updated named parameters and the mapping from parameter name to the shard function.
+    """
     bb = relax.BlockBuilder()
     param_to_shard_func = {}
     shard_func_names = set()
-    for name, param in model.state_dict().items():
+    new_named_params: Dict[str, nn.Parameter] = {}
+    has_shard_strategy = False
+    for name, param in named_params.items():
         shard_strategy = param.attrs.get("shard_strategy", None)
         if shard_strategy is not None:
-            _update_quantize_map(quantize_map, named_params, name, tensor_parallel_shards)
```

Comment:
need to keep original behavior because fp8 quantization depends on it

Reply:
Done. thanks for the catch!